src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java

```diff
@@ -11,59 +26,79 @@
+ *
+ * <p>The data contained in this sliding window is effectively a time series. Consumers may extend this
+ * class and define more complex aggregations or analytics functions.
+ */
+public class PersistableSlidingWindow extends SlidingWindow<SlidingWindowData> {
   private static final Logger LOG = LogManager.getLogger(PersistableSlidingWindow.class);
   private static final ObjectMapper objectMapper = new ObjectMapper();
 
   protected static final String SEPARATOR = "\n";
 
-  protected final long writePeriod;
+  protected final double bucketSizeInSeconds;
 
-  private Path path;
-  private BufferedWriter writer;
-  private long timestamp = -1;
+  private SlidingWindowData aggregatedData;
+  private Path pathToFile;
 
-  public PersistableSlidingWindow(int slidingWindowSize,
-                                  long writePeriod,
+  public PersistableSlidingWindow(int slidingWindowSizeInSeconds,
+                                  int bucketSizeInSeconds,
                                   TimeUnit timeUnit,
                                   String filePath) {
-    super(slidingWindowSize, timeUnit);
-    this.writePeriod = writePeriod;
+    super(slidingWindowSizeInSeconds, timeUnit);
+    this.bucketSizeInSeconds = bucketSizeInSeconds;
     try {
-      this.path = Paths.get(filePath);
-      if (Files.exists(path)) {
-        loadFromFile(path);
+      this.pathToFile = Paths.get(filePath);
+      if (Files.exists(pathToFile)) {
+        loadFromFile(pathToFile);
       } else {
-        Files.createFile(path);
+        Files.createFile(pathToFile);
       }
-      this.writer = new BufferedWriter(new FileWriter(filePath, false));
     } catch (IOException e) {
-      LOG.error("Couldn't create file {} to perform young generation tuning", path, e);
+      LOG.error("Couldn't create file {} to perform young generation tuning", pathToFile, e);
       throw new IllegalArgumentException("Couldn't create or read a file at " + filePath);
     }
   }
 
   @Override
   public void next(SlidingWindowData slidingWindowData) {
-    if (timestamp == -1) {
-      timestamp = slidingWindowData.getTimeStamp();
+    if (aggregatedData == null) {
+      aggregatedData = new SlidingWindowData(slidingWindowData.getTimeStamp(), slidingWindowData.getValue());
+    } else {
+      aggregatedData.setValue(aggregatedData.getValue() + slidingWindowData.getValue());
     }
     long currTimestamp = slidingWindowData.getTimeStamp();
-    super.next(slidingWindowData);
-    long timestampDiff = currTimestamp - timestamp;
-    if (timeUnit.convert(timestampDiff, TimeUnit.MILLISECONDS)
-        > timeUnit.convert(writePeriod, TimeUnit.MILLISECONDS)) {
+    long timestampDiff = currTimestamp - aggregatedData.getTimeStamp();
+    if (timeUnit.convert(timestampDiff, TimeUnit.SECONDS) > bucketSizeInSeconds) {
       try {
+        super.next(aggregatedData);
+        aggregatedData = null;
         persist();
```

Comment:
So we only persist when new buckets get added (at hour boundaries), not when individual bucket counters get updated?